ADBDEV-1153

tsdkv opened the pull request:
When the master host becomes unavailable, walreceiver is trying to open a connection with the master. After an unsuccessful attempt, it dies and reborns again. 

During the attempt, the system call poll() (or select) is called, where walreceiver is for a long time (on my cluster about 2 minutes).

When we call gpactivatestandby, the promote signal is sent. postmaster tries to stop the walreceiver and sends a `SIGTERM` to it. `SIGTERM` doesn't stop `poll()`. After the signals have arrived the `poll()` continues its work. You can check this by looking 
at the `pqSocketCheck` function in the `fe-misc.c` file. 

gpactivatestandby tries to connect to new master and by making 50 attempts in 50 seconds. If the connection fails, this error is displayed
```
[CRITICAL]:-Error activating standby master: Either the set port is incorrect or the postmaster could not come up.
```
Since `poll()` may run for several minutes, gpactivatestandby may never connect to the new master.
  
To fix this, we fixed the `libpqrcv_connect` function. The implementation logic is taken from newer versions of postgres.

Here's an stack trace for walreceiver's `poll()` call: 
```
#0  0x00007f0778556a20 in __poll_nocancel () from /lib64/libc.so.6
#1  0x000000000093957d in pqSocketPoll (end_time=-1, forWrite=0, forRead=0, sock=<optimized out>) at fe-misc.c:1236
#2  pqSocketCheck (conn=conn@entry=0x36a9590, forRead=forRead@entry=0, forWrite=forWrite@entry=1, end_time=end_time@entry=-1) at fe-misc.c:1178
#3  0x000000000093aa90 in pqWaitTimed (forRead=forRead@entry=0, forWrite=forWrite@entry=1, conn=conn@entry=0x36a9590, finish_time=finish_time@entry=-1)
    at fe-misc.c:1088
#4  0x000000000092ee35 in connectDBComplete (conn=0x36a9590) at fe-connect.c:1745
#5  0x000000000092fb4c in PQconnectdb (
    conninfo=conninfo@entry=0x7fff104d03e0 "user=gpadmin host=10.92.6.102 port=5432 sslmode=prefer sslcompression=1 krbsrvname=postgres application_name=gp_walreceiver dbname=replication replication=true fallback_application_name=walreceiver") at fe-connect.c:686
#6  0x0000000000a4abf6 in libpqrcv_connect (conninfo=<optimized out>) at libpqwalreceiver/libpqwalreceiver.c:109
#7  0x0000000000a40152 in WalReceiverMain () at walreceiver.c:326
#8  0x000000000079c792 in AuxiliaryProcessMain (argc=argc@entry=2, argv=argv@entry=0x7fff104d0e90) at bootstrap.c:459
#9  0x0000000000a1290d in StartChildProcess (type=WalReceiverProcess) at postmaster.c:5832
#10 MaybeStartWalReceiver () at postmaster.c:5987
#11 0x00000000006cc109 in ServerLoop () at postmaster.c:2017
#12 0x0000000000a15d17 in PostmasterMain (argc=argc@entry=6, argv=argv@entry=0x36870d0) at postmaster.c:1518
#13 0x00000000006d167b in main (argc=6, argv=0x36870d0) at main.c:245
```

leskin-in reviewed (approved):
I suppose `30` can be defined as some constant that depends on other values. I also believe it should be defined as a `static` variable. Maybe it can be made dependent on [`standby_message_timeout`](https://github.com/arenadata/gpdb/blob/82d8294baaa6a6122c371bf5f781b920cc32d341/src/bin/pg_basebackup/pg_basebackup.c#L69).

t1mursadykov commented:
Don't review this PR!

t1mursadykov commented:
Instead of timeout, it was decided to change the `libpqrcv_connect` function. The implementation logic is taken from newer versions of postgres.

maksm90 commented:
@t1mursadykov you could export some useful comments from poll loop in [master](https://github.com/postgres/postgres/blob/master/src/backend/replication/libpqwalreceiver/libpqwalreceiver.c#L172-L209) and [extra branch](https://github.com/postgres/postgres/blob/master/src/backend/replication/libpqwalreceiver/libpqwalreceiver.c#L185-L189) of connection status checking for Win32.
P.S. reset TAB expansion in your editor, your patch has indentation different from official code convention.

darthunix commented:
Should we extend behave tests for this case as well?

maksm90 commented:
@t1mursadykov fix compiler's issues:
```
extension.c:780:1: warning: ‘is_begin_state’ defined but not used [-Wunused-function]
 is_begin_state(const Node *stmt)
 ^~~~~~~~~~~~~~
scan.c: In function ‘yy_try_NUL_trans’:
scan.c:10189:23: warning: unused variable ‘yyg’ [-Wunused-variable]
     struct yyguts_t * yyg = (struct yyguts_t*)yyscanner; /* This var may be unused depending upon options. */
                       ^~~
In file included from bitmapinsert.c:24:0:
bitmapinsert.c: In function ‘inserttuple.constprop’:
../../../../src/include/access/bitmap.h:330:15: warning: array subscript is above array bounds [-Warray-bounds]
  (bool) ((((h)[(n)/BM_HRL_WORD_SIZE]) & (WORDNO_GET_HEADER_BIT(n))) > 0 ? \
           ~~~~^~~~~~~~~~~~~~~~~~~~~~~
bitmapinsert.c:602:8: note: in expansion of macro ‘IS_FILL_WORD’
    if (IS_FILL_WORD(new_words->hwords, wordNo))
        ^~~~~~~~~~~~
libpqwalreceiver/libpqwalreceiver.c: In function ‘libpqrcv_connect’:
libpqwalreceiver/libpqwalreceiver.c:121:2: error: ‘status’ undeclared (first use in this function)
  status = PGRES_POLLING_WRITING;
  ^~~~~~
libpqwalreceiver/libpqwalreceiver.c:121:2: note: each undeclared identifier is reported only once for each function it appears in
libpqwalreceiver/libpqwalreceiver.c:132:26: error: ‘MyLatch’ undeclared (first use in this function)
   rc = WaitLatchOrSocket(MyLatch,
                          ^~~~~~~
libpqwalreceiver/libpqwalreceiver.c:133:11: error: ‘WL_EXIT_ON_PM_DEATH’ undeclared (first use in this function)
           WL_EXIT_ON_PM_DEATH | WL_LATCH_SET | io_flag,
           ^~~~~~~~~~~~~~~~~~~
libpqwalreceiver/libpqwalreceiver.c:134:20: error: ‘conn’ undeclared (first use in this function)
           PQsocket(conn->streamConn),
                    ^~~~
make[3]: *** [libpqwalreceiver/libpqwalreceiver.o] Error 1
make[2]: *** [replication-recursive] Error 2
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [all-backend-recurse] Error 2
make: *** [all-src-recurse] Error 2
```

Commit:
Use asynchronous connection in libpqwalreceiver

The previous version, which used poll()/select(), caused some problems. For example, if the master is no longer available, we call gpactivatestandby while the walreceiver is trying to connect to the master. The walreceiver receives SIGTERM, but the poll() continues its work. gpactivatestandby tries to connect to new master, but it fails due to an expired timeout. To avoid such errors, WaitLatchOrSocket() is used, which is more portable and flexible.

Also change walreceiver to use its procLatch instead of a custom latch. The use of a walrcv->latch pointer is wrapped in a spinlock to prevent race condition.

maksm90 commented on src/backend/replication/libpqwalreceiver/libpqwalreceiver.c:
```diff
@@ -96,7 +96,7 @@ static void
 {
 	char		conninfo_repl[MAXCONNINFO + 75];
-
+	PostgresPollingStatusType status;
```
Put empty line after local variables declaration block

maksm90 commented on src/backend/replication/libpqwalreceiver/libpqwalreceiver.c:
```diff
@@ -106,13 +106,51 @@ libpqrcv_connect(char *conninfo)
 
-	streamConn = PQconnectdb(conninfo_repl);
+	streamConn =  PQconnectStart(conninfo_repl);
+
```
this empty line is redundant

maksm90 commented on src/backend/replication/libpqwalreceiver/libpqwalreceiver.c:
```diff
@@ -106,13 +107,50 @@ libpqrcv_connect(char *conninfo)
 				(errmsg("could not connect to the primary server: %s",
 						PQerrorMessage(streamConn))));
 }
-
```
This new line removal is redundant

Stolb27 commented on src/test/walrep/expected/walreceiver.out:
```diff
@@ -1,8 +1,8 @@
 -- negative cases
 SELECT test_receive();
-ERROR:  could not receive data from WAL stream: connection pointer is NULL (libpqwalreceiver.c:561)
+ERROR:  could not receive data from WAL stream: connection pointer is NULL (libpqwalreceiver.c:598)
```
Please point code correct line numbers here and below. Tests have failed:

```
 -- negative cases
 SELECT test_receive();
-ERROR:  could not receive data from WAL stream: connection pointer is NULL (libpqwalreceiver.c:598)
+ERROR:  could not receive data from WAL stream: connection pointer is NULL (libpqwalreceiver.c:600)
 SELECT test_send();
-ERROR:  could not send data to WAL stream: connection pointer is NULL (libpqwalreceiver.c:617)
+ERROR:  could not send data to WAL stream: connection pointer is NULL (libpqwalreceiver.c:619)
 SELECT test_disconnect();
  test_disconnect
```

Stolb27 reviewed (approved):
We have discussed changes in `WalRcvData` structure. We have API guarantees for extensions only between bugfix releases. So we will merge this fix and discuss it with Pivotal later.

darthunix commented:
AFAICS, a latch solution - is a back port of https://github.com/postgres/postgres/commit/597a87ccc9a6fa8af7f3cf280b1e24e41807d555 to GPDB. But later it was found out that we should protect this latch with a spin lock - it was fixed in https://github.com/postgres/postgres/commit/45f9d08684d954b0e514b69f270e763d2785dd53. @t1mursadykov do we need this commit as well?

t1mursadykov commented:
@darthunix There is a problem with porting  postgres/postgres@45f9d08, because we also use `WalRcv->latch` in signal handler. When the signal arrives, it could be a deadlock.

darthunix commented on src/backend/replication/libpqwalreceiver/libpqwalreceiver.c:
```diff
@@ -106,7 +107,45 @@ libpqrcv_connect(char *conninfo)
 
-	streamConn = PQconnectdb(conninfo_repl);
+	streamConn =  PQconnectStart(conninfo_repl);
+	if (PQstatus(streamConn) == CONNECTION_BAD)
```
We should add `streamConn == NULL` check as well.

tsdkv replied:
`PQstatus` returns `CONNECTION_BAD` when `streamConn` is `NULL`

darthunix reviewed (commented):
@t1mursadykov  `SetLatch(walrcv->latch)` is not only present in `WalRcvShutdownHandler()` but also in `RequestXLogStreaming()` as well and we can face latch tear down here too.

t1mursadykov commented:
Commit is ported and latch in signal handler is changed.

darthunix commented on src/backend/replication/walreceiver.c:
```diff
@@ -240,12 +241,13 @@ WalReceiverMain(void)
 	on_shmem_exit(WalRcvDie, 0);
 
-	OwnLatch(&walrcv->latch);
 
```
Redundant empty line

darthunix reviewed (changes requested):
I've thought a while about this PR and understood that we can't fix the problem in https://github.com/postgres/postgres/commit/45f9d08 with a simple https://github.com/arenadata/gpdb/pull/144/commits/96c4a2e901ac75cd27b324d5a85c5b6fbd5b7d32. Any read and write operations with a `latch` pointer should be wrapped with spin locks as they are not potentially atomic. Otherwise it is possible that our non-atomic modification of a latch pointer would be interrupted by a signal where we'll read not fully modified pointer and crash.

Also it was discussed that we can't use spin locks in a signal handler. But in fact everything is much worth: we can't use `ereport()` from a `SetLatch()` as well - look at https://github.com/postgres/postgres/commit/a1a789eb5ac894b4ca4b7742f2dc2d9602116e46. Though all of us are tied of this "small fix in gpactivatestandby" I suppose we need to think about porting this commit to a current PR.

**UPDATE.** I was wrong: https://github.com/postgres/postgres/commit/a1a789eb5ac894b4ca4b7742f2dc2d9602116e46 has already been backported and we really do protect ourselves from non-fully modified pointers with https://github.com/arenadata/gpdb/pull/144/commits/96c4a2e901ac75cd27b324d5a85c5b6fbd5b7d32 as `MyProc->procLatch` is never modified after signal handlers are set and we always read consistent data.

darthunix reviewed (approved):
I've inspected the code about latches (again, ha-ha...) and found out why we can't get rid of a shared latch (now @t1mursadykov initializes it with `MyProc->procLatch`). Latches can work as IPC as well and it is our case: startup process requests new XLOGs and uses shared `WalRcv` structure (including our shared latch) to send a signal with `SetLatch()` to the owner's process - walreciever. I think that a stack looks like
```
-- reading
StartupXLOG
ReadRecord
XLogReadRecord
ReadPageInternal (calls read_page())

-- initialization of the read_page() callback
StartupXLOG
XLogReaderAllocate
XLogPageRead
WaitForWALToBecomeAvailable
RequestXLogStreaming
SetLatch
```
When we call `SetLatch()` from a process that is not an owner of this latch (`owner_pid` field) it sends a SIGUSR1 signal to a target PID (walreciever). And walreciever processes this signal in it's signal handler and uses self-pipe trick to wake up `select()` in `WaitLatchOrSocket()`. Finally I got how it works and decided to share with others (but it seems that it was only me who didn't get the whole picture).

darthunix commented:
@t1mursadykov can you please open a discussion when you'll open this PR to upstream about incorrect backport of https://github.com/postgres/postgres/commit/a1a789eb5ac894b4ca4b7742f2dc2d9602116e46 to Greenplum 6X_STABLE (master is not affected). The problem is that [SetLatch() ](https://github.com/greenplum-db/gpdb/blob/cb2f8c0337640e918b646df25dc98a284531a7be/src/backend/replication/walreceiver.c#L795) in `WalRcvShutdownHandler` still has GP specific [elogif](https://github.com/greenplum-db/gpdb/blob/e2b0422bba7f0722294391b540a2d51fa1dc3cd5/src/backend/port/unix_latch.c#L527) print that can cause the same problem as backported commit tried to fix. On one hand it is a debug GUC for developers, but on the other it can cause troubles.
